bindings/rust/bench/certs/generate_certs.sh

```diff
@@ -6,36 +6,62 @@
-# security policy in s2n-tls.
-# https://github.com/aws/s2n-tls/blob/main/docs/USAGE-GUIDE.md#chart-security-policy-version-to-supported-curvesgroups
-echo "generating server private key and CSR"
-openssl req  -new -nodes -newkey ec -pkeyopt ec_paramgen_curve:secp384r1 -keyout server-key.pem -out server.csr -config config/server.cnf
+    # set openssl argument name
+    if [[ $1 == rsa ]]; then
+        local argname=rsa_keygen_bits:
+    elif [[ $1 == ec ]]; then
+        local argname=ec_paramgen_curve:P-
+    fi
 
-echo "generating client private key and CSR"
-openssl req  -new -nodes -newkey ec -pkeyopt ec_paramgen_curve:secp384r1 -keyout client-key.pem -out client.csr -config config/client.cnf
+    # make directory for certs
+    if [ ! -d $1$2/ ]; then
+        mkdir $1$2
+    fi
+    cd $1$2
 
-echo "generating server certificate and signing it"
-openssl x509 -days 65536 -req -in server.csr -CA ca-cert.pem -CAkey ca-key.pem -CAcreateserial -out server-cert.pem -extensions req_ext -extfile config/server.cnf
+    echo "generating CA private key and certificate"
+    openssl req -new -nodes -x509 -newkey $1 -pkeyopt $argname$2 -keyout  ca-key.pem -out ca-cert.pem -days 65536 -config ../config/ca.cnf
 
-echo "generating client certificate and signing it"
-openssl x509 -days 65536 -req -in client.csr -CA ca-cert.pem -CAkey ca-key.pem -CAcreateserial -out client-cert.pem -extensions req_ext -extfile config/client.cnf
+    echo "generating server private key and CSR"
+    openssl req  -new -nodes -newkey $1 -pkeyopt $argname$2 -keyout server-key.pem -out server.csr -config ../config/server.cnf
 
-echo "verifying generated certificates"
-openssl verify -CAfile ca-cert.pem server-cert.pem
-openssl verify -CAfile ca-cert.pem client-cert.pem
+    echo "generating client private key and CSR"
+    openssl req  -new -nodes -newkey $1 -pkeyopt $argname$2 -keyout client-key.pem -out client.csr -config ../config/client.cnf
 
-cat server-cert.pem ca-cert.pem > server-fullchain.pem
-cat client-cert.pem ca-cert.pem > client-fullchain.pem
+    echo "generating server certificate and signing it"
+    openssl x509 -days 65536 -req -in server.csr -CA ca-cert.pem -CAkey ca-key.pem -CAcreateserial -out server-cert.pem -extensions req_ext -extfile ../config/server.cnf
 
-echo "cleaning up temporary files"
-rm server.csr
-rm client.csr
-rm ca-key.pem
+    echo "generating client certificate and signing it"
+    openssl x509 -days 65536 -req -in client.csr -CA ca-cert.pem -CAkey ca-key.pem -CAcreateserial -out client-cert.pem -extensions req_ext -extfile ../config/client.cnf
+
+    echo "verifying generated certificates"
+    openssl verify -CAfile ca-cert.pem server-cert.pem
+    openssl verify -CAfile ca-cert.pem client-cert.pem
+
+    echo "combining certs into cert chains"
+    cat server-cert.pem ca-cert.pem > server-fullchain.pem
+    cat client-cert.pem ca-cert.pem > client-fullchain.pem
```

Comment:
To do auth, the server needs to send a certificate and the client needs a certificate that can verify the server's cert. Do you actually need to create the server-fullchain.pem file?

Reply:
All of the TLS libraries need to load both the leaf cert and root cert as part of a cert chain. I guess it's sort of a decision of whether or not the chain gets concatenated in the bash script or in the bench harness code, and concatenating them in the bash script just seems a lot easier.

Reply:
Definitely let's keep the pem logic here and not in the benchmarking code.
I'm not following why the chain is necessary. Shouldn't the client only need the root cert and the server only need the leaf cert? (Since the root cert verifies the leaf.) Like why would the server need to send the full chain of (root, leaf) if the client already has the root cert? Shouldn't it only need to send the leaf cert?

Reply:
I think it's probably because the client might have a lot of different root certs in it's root cert store, and having the root cert in the cert chain would tell the client reading the cert chain which root cert is being used. Also, if the chain is longer (which it almost always is in a real-world use case), you would need to include the whole chain and not just to leaf cert to link the leaf back to the root.

Reply:
I do agree that certificate chains are longer in the real world. But the server wouldn't send information that the client already has, that is really expensive. Are you getting an error if you don't use the full chain?

Reply:
Oh I don't get an error, it does work, that's great! I'll take it out.